application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml

```diff
@@ -0,0 +1,195 @@
+      #set ($query = $query + ' and task.duedate is not NULL')
+    #end
+    #if ("$!request.to" != '')
+      #set ($to = $datetool.toDate($numbertool.toNumber($request.to).longValue()))
+      #set ($query = $query + ' and (' +
+              '(task.duedate is not NULL and task.duedate &lt;= :to) or ' +
+              '(task.duedate is     NULL and (' +
+                '(task.startDate is not NULL and task.startDate  &lt;= :to) or ' +
+                '(task.startDate is     NULL and task.createDate &lt;= :to))' +
+            '))'
+      )
+    #end
+    ## Sort the tasks by project and name to ensure a stable order.
+    #set ($query = $query + ' order by UPPER(task.project), task.project, UPPER(task.name), task.name')
+    #set ($query = $services.query.xwql($query).setLimit(50))
+    #if ($spaces)
+      #set ($query = $query.bindValue("spaces", $spaces))
+      #set ($index = 0)
+      #foreach ($space in $spaces)
+        #set ($space = $space + '.%')
+        #set ($query = $query.bindValue("spaces${index}", $space))
+        #set ($index = $index + 1)
+      #end
+    #end
+    #if ($assignees) #set ($query = $query.bindValue('assignees', $assignees)) #end
+    #if ($projects)  #set ($query = $query.bindValue('projects', $projects))   #end
+    #if ($reporters) #set ($query = $query.bindValue('reporters', $reporters)) #end
+    #if ($from)      #set ($query = $query.bindValue('from', $from))           #end
+    #if ($to)        #set ($query = $query.bindValue('to', $to))               #end
+  #end
+  #set ($queryResult = $query.execute())
+
+  #set ($tasks = [])
+  #foreach ($taskDocument in $queryResult)
+    #set ($taskXDoc = $xwiki.getDocument($taskDocument))
+    #set ($task = $taskXDoc.getObject('TaskManager.TaskManagerClass'))
+    #if ($services.security.authorization.hasAccess('view', $taskDocument))
+      #set ($taskObj = {})
+      #set ($customClasses = '')
+      #set ($discard = $taskObj.put('id', "$taskXDoc.getDocumentReference()"))
+      #set ($discard = $taskObj.put('link', $taskXDoc.getURL()))
+      #putOrDefault ($taskObj 'actualName'
+        $task.getValue('name')
+        $services.localization.render('taskmanager.gantt.task.unknownTaskTitle')
+      )
+      #putOrDefault ($taskObj 'progress'
+        $task.getValue('progress')
+        0
+      )
+      #putOrDefault ($taskObj 'project'  $task.getValue('project')  '')
+      #putOrDefault ($taskObj 'assignee' $task.getValue('assignee') '')
+      #putOrDefault ($taskObj 'reporter' $task.getValue('reporter') '')
+      #putOrDefault ($taskObj 'start'
+        $task.getValue('startDate').getTime()
+        $task.getValue('createDate').getTime()
+      )
+      #putOrDefault ($taskObj 'end'
+        $task.getValue('duedate').getTime()
+        $xwiki.jodatime.getDateTime($taskObj['start']).plusWeeks(1).getMillis()
+      )
+      #if ("$!task.getValue('duedate').getTime()" == '')
+        #set ($customClasses = $customClasses + ' taskgantt-no-due-date')
```

Comment:
Maybe you could have a comma separated list of custom classes. I think it will be more intuitive to split then in JS by `,` instead of space.
Here, in velocity, you will have a list `#set ($customClasses = [])` where it's easier to add values than concatenating values. At the end you can use `$stringtool.join($customClasses, ',')`

I think this will also remove the need for `$task.getValue('status').replace(' ', '-')` below, right?

Reply:
It can be changed, but the ```$task.getValue('status').replace(' ', '-')``` would still need to be there, either in velocity or JS.

Why the replace is needed: The ```custom_class``` parameter in the Gantt library is directly inserted as the class attribute in the task bar. So if a status has a space, its associated class would be interpreted as 2 different classes (Ex: status "To Do" -> class "taskgantt-status-To Do", which means the element will have css classes "taskgantt-status-To" and "Do").

Reply:
To ask Stefana is we need the custom classes for toDo, inProgress and Done